File: app/src/main/java/com/princeparadoxes/watertracker/presentation/screen/statistic/StatisticFragment.java
package com.princeparadoxes.watertracker.presentation.screen.statistic;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomSheetBehavior;
import android.text.SpannableStringBuilder;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.TextView;

import com.daimajia.swipe.SwipeLayout;
import com.princeparadoxes.watertracker.ProjectApplication;
import com.princeparadoxes.watertracker.R;
import com.princeparadoxes.watertracker.base.BaseFragment;
import com.princeparadoxes.watertracker.data.repository.DrinkRepository;
import com.princeparadoxes.watertracker.data.source.sp.ProjectPreferences;
import com.princeparadoxes.watertracker.utils.AnimatorUtils;
import com.princeparadoxes.watertracker.utils.DimenTools;
import com.princeparadoxes.watertracker.utils.SpannableUtils;
import com.yarolegovich.discretescrollview.DiscreteScrollView;
import com.yarolegovich.discretescrollview.transform.ScaleTransformer;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindDrawable;
import butterknife.BindView;
import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

public class StatisticFragment extends BaseFragment
        implements
        DiscreteScrollView.OnItemChangedListener<StatisticItemViewHolder>,
        DiscreteScrollView.ScrollStateChangeListener<StatisticItemViewHolder> {

    ///////////////////////////////////////////////////////////////////////////
    // INJECTS
    ///////////////////////////////////////////////////////////////////////////
    @Inject
    StatisticViewModel statisticViewModel;

    ///////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////  VIEWS  //////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////////////////////

    @BindView(R.id.statistic_header)
    SwipeLayout mHeaderView;
    @BindView(R.id.statistic_header_start)
    TextView mStatisticHeaderStart;
    @BindView(R.id.statistic_header_center)
    TextView mStatisticHeaderCenter;
    @BindView(R.id.statistic_header_end)
    TextView mStatisticHeaderEnd;
    @BindView(R.id.statistic_type_view)
    DiscreteScrollView mTypePicker;
    @BindView(R.id.statistic_forecast_view)
    StatisticChartView mStatisticChartView;
    @BindView(R.id.statistic_top_border)
    View mTopBorderView;

    @BindDrawable(R.drawable.ic_chevron_down)
    Drawable mChevronDownDrawable;
    @BindDrawable(R.drawable.ic_chevron_up)
    Drawable mChevronUpDrawable;

    ///////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////  FIELDS  /////////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////////////////////

    private BottomSheetBehavior mStatisticBottomSheetBehavior;
    private StatisticAdapter mStatisticAdapter;

    ///////////////////////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////  INIT SCREEN  /////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ProjectApplication app = ProjectApplication.get(getActivity());
        StatisticScope.Component component = DaggerStatisticScope_Component.builder()
                .projectComponent(app.component())
                .module(new StatisticScope.Module(this))
                .build();
        component.inject(this);
    }

    @Override
    protected int layoutId() {
        return R.layout.fragment_statistic;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        mStatisticAdapter = new StatisticAdapter();
        initHeader();
        initBehavior(container);
        onCollapsed();
        return view;
    }

    ///////////////////////////////////////////////////////////////////////////
    // HEADER
    ///////////////////////////////////////////////////////////////////////////

    private void initHeader() {
        mHeaderView.setLeftSwipeEnabled(true);
        mHeaderView.setRightSwipeEnabled(true);
    }

    private void changeHeaderState(int newState) {
        switch (newState) {
            case BottomSheetBehavior.STATE_COLLAPSED:
                onCollapsed();
                break;
            case BottomSheetBehavior.STATE_EXPANDED:
                onExpanded();
                break;
        }
    }

    private void onCollapsed() {
        mTopBorderView.setVisibility(View.VISIBLE);
        createHeaderAnimator(R.string.statistic_header_open, mChevronUpDrawable).start();
    }

    private void onExpanded() {
        mTopBorderView.setVisibility(View.INVISIBLE);
        createHeaderAnimator(R.string.statistic_header_closed, mChevronDownDrawable).start();
//        SpannableStringBuilder builder = new SpannableStringBuilder();
//        builder.append(SpannableUtils.getAbsoluteSizeSpan(getContext(), "Day norm", 12))
//                .append("\n")
//                .append(String.valueOf(mProjectPreferences.getCurrentDayNorm()))
//                .append("ml");
//
//        mStatisticHeaderStart.setText(builder);
//        mStatisticHeaderEnd.setText("Send feedback");
    }

    @NonNull
    private AnimatorSet createHeaderAnimator(int text, Drawable drawable) {
        AnimatorSet set = new AnimatorSet();
        Animator out = AnimatorUtils.alphaAnimator(1, 0, mStatisticHeaderCenter, 200, new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                mStatisticHeaderCenter.setText(text);
                mStatisticHeaderCenter.setCompoundDrawablesWithIntrinsicBounds(null, drawable, null, null);
            }
        });
        out.setInterpolator(new AccelerateInterpolator());
        Animator in = AnimatorUtils.alphaAnimator(0, 1, mStatisticHeaderCenter, 200);
        in.setInterpolator(new DecelerateInterpolator());
        set.playSequentially(out, in);
        return set;
    }

    ///////////////////////////////////////////////////////////////////////////
    // BEHAVIOR
    ///////////////////////////////////////////////////////////////////////////

    private void initBehavior(ViewGroup container) {
        mStatisticBottomSheetBehavior = BottomSheetBehavior.from(container);
        mStatisticBottomSheetBehavior.setBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {
                changeHeaderState(newState);
            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {

            }
        });
        mStatisticBottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
        mStatisticBottomSheetBehavior.setHideable(false);
        mStatisticBottomSheetBehavior.setPeekHeight((int) DimenTools.pxFromDp(getContext(), 64));
    }


    ///////////////////////////////////////////////////////////////////////////////////////////////
    ///////////////////////////////////  START SCREEN  ////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onStart() {
        super.onStart();
        initTypePicker();
        unsubscribeOnStop(
                statisticViewModel.getStatistic().subscribe(this::handleLoadAllStatistic, Timber::e),
                statisticViewModel.getDayNorm().subscribe(this::handleLoadDaySum, Timber::e),
                statisticViewModel.getLast().subscribe(this::handleLoadLastDrink, Timber::e)
        );
    }

    private void initTypePicker() {
        mTypePicker.setAdapter(mStatisticAdapter);
        mTypePicker.setOnItemChangedListener(this);
        mTypePicker.setScrollStateChangeListener(this);
        mTypePicker.scrollToPosition(0);
        mTypePicker.setItemTransitionTimeMillis(150);
        mTypePicker.setItemTransformer(new ScaleTransformer.Builder()
                .setMinScale(0.8f)
                .build());
    }

    private void handleLoadAllStatistic(List<StatisticModel> statisticModels) {
        mStatisticAdapter.setData(statisticModels);
    }

    private void handleLoadDaySum(int daySum) {
        SpannableStringBuilder builder = new SpannableStringBuilder();
        builder.append(SpannableUtils.getAbsoluteSizeSpan(getContext(), "Day", 12))
                .append("\n")
                .append(String.valueOf(daySum))
                .append("ml");
        mStatisticHeaderStart.setText(builder);
    }

    private void handleLoadLastDrink(int lastByDay) {
        SpannableStringBuilder builder = new SpannableStringBuilder();
        builder.append(SpannableUtils.getAbsoluteSizeSpan(getContext(), "Last", 12))
                .append("\n")
                .append(lastByDay != 0 ? String.valueOf(lastByDay) + "ml" : "none");
        mStatisticHeaderEnd.setText(builder);
    }

    ///////////////////////////////////////////////////////////////////////////
    // TYPE PICKER
    ///////////////////////////////////////////////////////////////////////////

    @Override
    public void onCurrentItemChanged(@NonNull StatisticItemViewHolder viewHolder, int adapterPosition) {
        mStatisticChartView.bindView(
                new float[]{0f, 2f, 1.4f, 4.f, 3.5f, 4.3f, 2f, 4f, 6.f},
                new float[]{0f, 2f, 1.4f, 4.f, 3.5f, 4.3f, 2f, 4f, 6.f});
        viewHolder.onChanged();
    }

    @Override
    public void onScrollStart(@NonNull StatisticItemViewHolder currentItemHolder, int adapterPosition) {
        currentItemHolder.onUnchanged();
    }

    @Override
    public void onScrollEnd(@NonNull StatisticItemViewHolder currentItemHolder, int adapterPosition) {

    }

    @Override
    public void onScroll(float scrollPosition,
                         @NonNull StatisticItemViewHolder currentHolder,
                         @NonNull StatisticItemViewHolder newCurrent) {
//        StatisticType current = mStatisticTypes.get(mTypePicker.getCurrentItem());
//        int nextPosition = mTypePicker.getCurrentItem() + (scrollPosition > 0 ? -1 : 1);
//        if (nextPosition >= 0 && nextPosition < mTypePicker.getAdapter().getItemCount()) {
//            StatisticType next = mStatisticTypes.get(nextPosition);
//            mStatisticChartView.onScroll(1f - Math.abs(scrollPosition), current, next);
//        }
    }

    @Override
    public void onResume() {
        super.onResume();

    }

    ///////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////  STOP SCREEN  ////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mStatisticBottomSheetBehavior.setBottomSheetCallback(null);
    }

    ///////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////  INSTANCE  ///////////////////////////////////////////////
    ///////////////////////////////////////////////////////////////////////////////////////////////

    public static StatisticFragment newInstance() {
        Bundle args = new Bundle();
        StatisticFragment fragment = new StatisticFragment();
        fragment.setArguments(args);
        return fragment;
    }

}

File: app/src/main/java/com/princeparadoxes/watertracker/presentation/screen/statistic/StatisticFragment.kt
package com.princeparadoxes.watertracker.presentation.screen.statistic

import android.animation.Animator
import android.animation.ObjectAnimator
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.design.widget.BottomSheetBehavior
import android.support.v4.content.ContextCompat
import android.text.SpannableStringBuilder
import android.transition.Fade
import android.view.View
import android.view.ViewGroup
import android.view.animation.BounceInterpolator
import android.widget.ImageView
import android.widget.TextView
import com.daimajia.swipe.SwipeLayout
import com.jakewharton.rxbinding2.view.RxView
import com.princeparadoxes.watertracker.ProjectApplication
import com.princeparadoxes.watertracker.R
import com.princeparadoxes.watertracker.presentation.base.BaseFragment
import com.princeparadoxes.watertracker.domain.entity.Drink
import com.princeparadoxes.watertracker.domain.entity.StatisticModel
import com.princeparadoxes.watertracker.presentation.base.FragmentSwitcherCompat
import com.princeparadoxes.watertracker.presentation.screen.settings.SettingsFragment
import com.princeparadoxes.watertracker.presentation.utils.ApplicationSwitcher
import com.princeparadoxes.watertracker.presentation.utils.DimenTools
import com.princeparadoxes.watertracker.presentation.utils.SpannableUtils
import com.princeparadoxes.watertracker.safeSubscribe
import com.yarolegovich.discretescrollview.DiscreteScrollView
import com.yarolegovich.discretescrollview.transform.ScaleTransformer
import javax.inject.Inject

class StatisticFragment : BaseFragment(), DiscreteScrollView.OnItemChangedListener<StatisticItemViewHolder>, DiscreteScrollView.ScrollStateChangeListener<StatisticItemViewHolder> {

    ///////////////////////////////////////////////////////////////////////////
    // INJECTS
    ///////////////////////////////////////////////////////////////////////////

    @Inject
    internal lateinit var statisticViewModel: StatisticViewModel

    ///////////////////////////////////////////////////////////////////////////
    // RESOURCES
    ///////////////////////////////////////////////////////////////////////////

    private val headerView by lazy { view!!.findViewById(R.id.statistic_header) as SwipeLayout }
    private val headerSettings by lazy { view!!.findViewById(R.id.statistic_header_settings) as TextView }
    private val headerReport by lazy { view!!.findViewById(R.id.statistic_header_report) as TextView }
    private val headerRevert by lazy { view!!.findViewById(R.id.statistic_header_revert) as TextView }
    private val headerStartText by lazy { view!!.findViewById(R.id.statistic_header_start) as TextView }
    private val headerCenterText by lazy { view!!.findViewById(R.id.statistic_header_center) as TextView }
    private val headerCenterTextImage by lazy { view!!.findViewById(R.id.statistic_header_center_image) as ImageView }
    private val headerEndText by lazy { view!!.findViewById(R.id.statistic_header_end) as TextView }
    private val typePicker by lazy { view!!.findViewById(R.id.statistic_type_view) as DiscreteScrollView }
    private val chartView by lazy { view!!.findViewById(R.id.statistic_forecast_view) as StatisticChartView }
    private val topBorderView by lazy { view!!.findViewById(R.id.statistic_top_border) as View }

    private val chevronDownDrawable by lazy { ContextCompat.getDrawable(context!!, R.drawable.ic_chevron_down) }
    private val chevronUpDrawable by lazy { ContextCompat.getDrawable(context!!, R.drawable.ic_chevron_up) }

    ///////////////////////////////////////////////////////////////////////////
    // FIELDS
    ///////////////////////////////////////////////////////////////////////////

    private val bottomSheetBehavior by lazy { BottomSheetBehavior.from(view!!.parent as ViewGroup)!! }
    private val statisticAdapter by lazy { StatisticAdapter() }

    ///////////////////////////////////////////////////////////////////////////
    // INIT SCREEN
    ///////////////////////////////////////////////////////////////////////////

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        DaggerStatisticScope_Component.builder()
                .projectComponent(ProjectApplication.component())
                .module(StatisticScope.Module(this))
                .build()
                .inject(this)
    }

    override fun layoutId(): Int = R.layout.fragment_statistic

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initHeader()
        initBehavior(view.parent as ViewGroup)
        onCollapsed()
        initTypePicker()
    }

    private fun initTypePicker() {
        typePicker.apply {
            adapter = statisticAdapter
            setOnItemChangedListener(this@StatisticFragment)
            setScrollStateChangeListener(this@StatisticFragment)
            scrollToPosition(0)
            setItemTransitionTimeMillis(150)
            setItemTransformer(ScaleTransformer.Builder().setMinScale(0.8f).build())
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // HEADER
    ///////////////////////////////////////////////////////////////////////////

    private fun initHeader() {
        headerView.apply {
            isLeftSwipeEnabled = true
            isRightSwipeEnabled = true
        }
    }

    private fun changeHeaderState(newState: Int) {
        when (newState) {
            BottomSheetBehavior.STATE_COLLAPSED -> onCollapsed()
            BottomSheetBehavior.STATE_EXPANDED -> onExpanded()
        }
    }

    private fun onCollapsed() {
        topBorderView.visibility = View.VISIBLE
        createHeaderAnimator(R.string.statistic_header_open, chevronUpDrawable).start()
    }

    private fun onExpanded() {
        topBorderView.visibility = View.INVISIBLE
        createHeaderAnimator(R.string.statistic_header_closed, chevronDownDrawable).start()
    }

    private fun createHeaderAnimator(text: Int, drawable: Drawable?): Animator {
        return when (Math.round(headerCenterTextImage.rotation) == 180) {
            true -> ObjectAnimator.ofFloat(headerCenterTextImage, View.ROTATION, 180F, 0F)
            false -> ObjectAnimator.ofFloat(headerCenterTextImage, View.ROTATION, 0F, 180F)
        }.apply {
            duration = 600
            interpolator = BounceInterpolator()
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // BEHAVIOR
    ///////////////////////////////////////////////////////////////////////////

    private fun initBehavior(container: ViewGroup?) {
        bottomSheetBehavior.setBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                changeHeaderState(newState)
            }

            override fun onSlide(bottomSheet: View, slideOffset: Float) {

            }
        })
        bottomSheetBehavior.apply {
            state = BottomSheetBehavior.STATE_COLLAPSED
            isHideable = false
            peekHeight = DimenTools.pxFromDp(context, 64f).toInt()
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // START SCREEN
    ///////////////////////////////////////////////////////////////////////////

    override fun onStart() {
        super.onStart()
        unsubscribeOnStop(
                statisticViewModel.observeDeleteWater(RxView.clicks(headerRevert)).safeSubscribe { handleDeleteWater(it) },
                statisticViewModel.observeReport(RxView.clicks(headerReport)).safeSubscribe { handleReport() },
                statisticViewModel.observeSetting(RxView.clicks(headerSettings)).safeSubscribe {
                    FragmentSwitcherCompat.start(fragmentManager)
                            .fragment(SettingsFragment.newInstance())
                            .containerId(R.id.main_start_fragment_container)
                            .replace();
                    BottomSheetBehavior.from<View>(activity!!.findViewById(R.id.main_start_fragment_container)).state = BottomSheetBehavior.STATE_EXPANDED
                },
                statisticViewModel.observeStatistic().safeSubscribe({ this.handleStatistic(it) }),
                statisticViewModel.observeDaySum().safeSubscribe({ this.handleDaySum(it) }),
                statisticViewModel.observeLastDrink().safeSubscribe({ this.handleLastDrink(it) }),
                statisticViewModel.observeDrinksByPeriod().safeSubscribe({ this.handleDrinksByPeriod(it) }),
                statisticViewModel.observeDetailStatistic().safeSubscribe({ this.handleAverage(it) })
        )
    }

    private fun handleReport() {
        ApplicationSwitcher.start(context).openEmailApplication("", "I have a problem")
    }

    private fun handleDeleteWater(it: Int) {
    }

    private fun handleAverage(it: List<Int>) {
        chartView.bindView(
                it.mapIndexed { index, _ -> index.toFloat() }.toFloatArray(),
                it.map { it.toFloat() }.toFloatArray())
    }

    private fun handleStatistic(statisticModels: List<StatisticModel>) {
        statisticAdapter.setData(statisticModels)
    }

    private fun handleDaySum(daySum: Int) {
        val builder = SpannableStringBuilder()
        builder.append(SpannableUtils.getAbsoluteSizeSpan(context, "Day", 12))
                .append("\n")
                .append(daySum.toString())
                .append("ml")
        headerStartText.text = builder
    }

    private fun handleLastDrink(lastByDay: Int) {
        val builder = SpannableStringBuilder()
        builder.append(SpannableUtils.getAbsoluteSizeSpan(context, "Last", 12))
                .append("\n")
                .append(if (lastByDay != 0) lastByDay.toString() + "ml" else "none")
        headerEndText.text = builder
    }

    private fun handleDrinksByPeriod(drinks: List<Drink>) {
        if (drinks.isEmpty()) return
//        chartView.bindView(
//                drinks.mapIndexed { index, drink -> index.toFloat() }.toFloatArray(),
//                drinks.map { it.size.toFloat() }.toFloatArray())
    }

    ///////////////////////////////////////////////////////////////////////////
    // TYPE PICKER
    ///////////////////////////////////////////////////////////////////////////

    override fun onCurrentItemChanged(viewHolder: StatisticItemViewHolder, adapterPosition: Int) {
        statisticViewModel.changeDrinkPeriod(statisticAdapter.getItem(adapterPosition).statisticType)
        viewHolder.onChanged()
    }

    override fun onScrollStart(currentItemHolder: StatisticItemViewHolder, adapterPosition: Int) {
        currentItemHolder.onUnchanged()
    }

    override fun onScrollEnd(currentItemHolder: StatisticItemViewHolder, adapterPosition: Int) {
    }

    override fun onScroll(scrollPosition: Float,
                          currentHolder: StatisticItemViewHolder,
                          newCurrent: StatisticItemViewHolder) {
        //        StatisticType current = mStatisticTypes.get(typePicker.getCurrentItem());
        //        int nextPosition = typePicker.getCurrentItem() + (scrollPosition > 0 ? -1 : 1);
        //        if (nextPosition >= 0 && nextPosition < typePicker.getAdapter().getItemCount()) {
        //            StatisticType next = mStatisticTypes.get(nextPosition);
        //            chartView.onScroll(1f - Math.abs(scrollPosition), current, next);
        //        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // STOP SCREEN
    ///////////////////////////////////////////////////////////////////////////

    override fun onDestroyView() {
        super.onDestroyView()
        bottomSheetBehavior.setBottomSheetCallback(null)
    }

    companion object {

        fun newInstance(): StatisticFragment {
            val args = Bundle()
            val fragment = StatisticFragment()
            fragment.arguments = args
            return fragment
        }
    }

}
